tasks/parameter/sp.yml

```diff
@@ -21,8 +21,8 @@
 
 - name: Set service principal parameter
   set_fact:
-      sp: "{{ sp | default({}) | combine({ item.key: vars[item.value] }) }}"
+      sp: "{{ sp | default({}) | combine({ item.key: lookup('vars', item.value) }) }}"
   with_items:
       - {'key': 'client_id', 'value': 'service_principal' }
       - {'key': 'client_secret', 'value': 'client_secret' }
-  when: service_principal is defined and {{ item.value }} is defined
+  when: service_principal is defined and vars[item.value] is defined
```

Comment:
I replaced `{{ item.value }}` with the `vars` keyword to resolve the variable (literal is fine here)